rush-state/src/console.rs

```diff
@@ -128,72 +130,43 @@ pub fn restore_terminal() {
     RAW_MODE.store(false, Ordering::Release);
 }
 
-impl<'a> Console<'a> {
-    pub fn new() -> Result<Self> {
-        let backend = CrosstermBackend::new(stdout());
-        let terminal = Terminal::new(backend)?;
-
-        Ok(Self {
-            terminal,
-            data: ConsoleData::new(),
-        })
-    }
-
-    // Enters the TUI console
-    pub fn enter(&mut self) -> Result<()> {
-        enable_raw_mode()?;
-        // ? Is mouse capture enabled by default?
-        execute!(
-            self.terminal.backend_mut(),
-            EnterAlternateScreen,
-            DisableMouseCapture
-        )?;
-        self.terminal.show_cursor()?;
-
-        RAW_MODE.store(true, Ordering::Release);
-        self.clear(ClearMode::LINE)
-    }
-
-    // Closes the TUI console and exits the program
-    // * Error handling here is unnecessary because the program is exiting
-    // TODO: This assumption may need to be reevaluated in the future
-    pub fn exit(&mut self, code: i32) {
-        restore_terminal();
-        // TODO: Exiting ignores drops, which may be problematic
-        std::process::exit(code);
-    }
-
+impl Console<'static> {
     // Reads a line of input from the user
     // Handles all TUI interaction between the user and the prompt
-    pub fn read_line(&mut self, shell: &Shell) -> Result<String> {
-        self.data.update_output_tick(shell);
-        self.data.update_prompt(shell);
-        self.data.update_debug(shell);
-        self.draw_frame(true)?;
+    pub fn read_line(console_ref: &Arc<RwLock<Self>>, shell: &mut Shell) -> Result<String> {
+        {
+            let mut console = console_ref.write().unwrap();
+            console.data.update_output_tick(shell);
+            console.data.update_prompt(shell);
+            console.data.update_debug(shell);
+            console.draw_frame(true)?;
+        }
 
         loop {
+            // console is unlocked while blocking here to allow the autocomplete to update
```

Comment:
(For all comments) make sure to capitalize the first letter. And yes this is just my OCD talking.